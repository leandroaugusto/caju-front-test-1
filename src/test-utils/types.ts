export interface ITestUtils {
  children?: React.ReactNode
}
