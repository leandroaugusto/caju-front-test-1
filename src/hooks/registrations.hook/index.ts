import axios from "axios";
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";

import type { TRegistrationsData } from "~/types/registrations.types";

async function fetchRegistrations() {
  const { data } = await axios.get("http://localhost:3000/registrations");
  return data;
}

async function saveRegistration(params: Partial<TRegistrationsData>) {
  const { data } = await axios.post("http://localhost:3000/registrations", params);
  return data;
}

export function useRegistrationsHook() {
  const queryClient = useQueryClient();

  const fetch = useQuery({
    queryKey: ["registrations"],
    queryFn: fetchRegistrations,
  });

  const mutation = useMutation({
    mutationFn: saveRegistration,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["registrations"] })
    },
  })

  return { fetch, mutation }
}
