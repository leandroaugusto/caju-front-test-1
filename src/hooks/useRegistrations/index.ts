import { useQuery, useMutation, useQueryClient, keepPreviousData, DefaultError } from "@tanstack/react-query";

import { TRegistrationsData } from "~/types/registrations.types";

import * as Services from "~/services";

import { queryKey } from "~/constants";

export const useFetchRegistrationsByCpfHook = (cpf: string) => {
  const queryClient = useQueryClient();

  return useQuery({
    queryKey: [queryKey, cpf],
    queryFn: async () => {
      const cachedData: TRegistrationsData[] | undefined = queryClient.getQueryData([queryKey]);

      if (cachedData) {
        const data = cachedData?.filter((user) => user.cpf === cpf);
        if (data.length) return data;
      }

      const fetchedData = await Services.fetchRegistrationByCPF(cpf);

      queryClient.setQueryData([queryKey], (oldData: TRegistrationsData[]) => {
        return oldData ? [...oldData, fetchedData] : [fetchedData];
      });

      return fetchedData;
    },
    enabled: !!cpf
  });
};

export function useFetchAllRegistrationsHook() {
  return useQuery({
    queryKey: [queryKey],
    queryFn: () => Services.fetchAllRegistrations(),
    enabled: true,
    placeholderData: keepPreviousData,
    refetchOnMount: false,
    refetchOnWindowFocus: false,
  });
}

export function useRegistrationsHook() {
  const queryClient = useQueryClient();

  const saveUser = useMutation({
    mutationFn: Services.saveRegistration,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: [queryKey] })
    },
  });

  const deleteUser = useMutation({
    mutationFn: Services.deleteRegistration,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: [queryKey] })
    },
  })

  const reviewUser = useMutation({
    mutationFn: Services.reviewRegistration,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: [queryKey] })
    },
  })

  return { saveUser, deleteUser, reviewUser }
}
