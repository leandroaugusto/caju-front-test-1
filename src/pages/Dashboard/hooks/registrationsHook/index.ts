import { useQuery } from "@tanstack/react-query";

async function fetchRegistrations() {
  const response = await fetch("http://localhost:3000/registrations/");
  const data = await response.json();
  return data;
}

export function useRegistrationsHook() {
  return useQuery({
    queryKey: ["registrations"],
    queryFn: fetchRegistrations,
  });
}
