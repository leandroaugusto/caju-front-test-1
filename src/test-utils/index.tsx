import { ReactElement } from "react";
import { ThemeProvider } from "styled-components";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import { RegistrationsContextProvider } from "~/contexts/registrations.context";
import { ITestUtils } from "./types";

import theme from "~/theme";

const queryClient = new QueryClient();

const AllTheProviders: React.FC = ({ children }: ITestUtils) => {
  return (
    <QueryClientProvider client={queryClient}>
      <RegistrationsContextProvider>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </RegistrationsContextProvider>
    </QueryClientProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult =>
  render(ui, {
    wrapper: AllTheProviders,
    ...options,
  });

export * from "@testing-library/react";

export { customRender };
