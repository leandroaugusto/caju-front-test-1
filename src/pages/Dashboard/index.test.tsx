import { renderHook } from "@testing-library/react-hooks";

import { customRender, screen, waitFor } from "~/test-utils";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import {
  ERegistrationsStatus,
  TRegistrationsData,
} from "~/types/registrations.types";
import { useRegistrationsHook } from "~/hooks/useRegistrations";

import DashboardPage from ".";

const mockRegistrations: TRegistrationsData[] = [
  {
    admissionDate: "22/10/2023",
    email: "luiz@caju.com.br",
    employeeName: "Luiz Filho",
    status: ERegistrationsStatus.APPROVED,
    cpf: "56642105087",
    id: "3",
  },
  {
    admissionDate: "22/10/2023",
    email: "luiz@caju.com",
    employeeName: "Luiz Neto",
    status: ERegistrationsStatus.REVIEW,
    cpf: "56642105088",
    id: "2",
  },
];

const mockUseQuery = jest.fn(() => ({
  isLoading: true,
  isSuccess: false,
  error: null,
  data: null,
})) as jest.Mock;

const queryClient = new QueryClient();
const wrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);

jest.mock("@tanstack/react-query", () => ({
  ...jest.requireActual("@tanstack/react-query"),
  useQuery: () => mockUseQuery(),
}));

jest.mock("./components/Searchbar", () => ({
  SearchBar: () => <div data-testid="mock-search-bar-container">SearchBar</div>,
}));

jest.mock("./components/Columns", () => ({
  Columns: ({ registrations }: { registrations: TRegistrationsData[] }) => (
    <div data-testid="mock-columns-container">
      {registrations.map((i) => (
        <div key={i.id}>{i.id}</div>
      ))}
    </div>
  ),
}));

describe("DashboardPage", () => {
  it("Should show DashboardPage page", async () => {
    mockUseQuery.mockImplementation(() => ({
      isLoading: false,
      isSuccess: true,
      error: null,
      data: mockRegistrations,
    }));

    const { result } = renderHook(() => useRegistrationsHook(), { wrapper });
    await waitFor(() => expect(result.current.fetch.isSuccess).toBe(true));

    const { container } = customRender(<DashboardPage />);

    const mockSearchBarContainer = screen.getByTestId(
      "mock-search-bar-container"
    );
    const mockColumnsContainer = screen.getByTestId("mock-columns-container");

    expect(mockSearchBarContainer).toBeInTheDocument();
    expect(mockColumnsContainer).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it("Should render loading state when data is not available", async () => {
    mockUseQuery.mockImplementation(() => ({
      ...mockUseQuery,
      isLoading: true,
    }));

    const { result } = renderHook(() => useRegistrationsHook(), { wrapper });
    await waitFor(() => expect(result.current.fetch.isLoading).toBe(true));

    customRender(<DashboardPage />);

    const mockLoadingContainer = screen.getByTestId("loading-container");

    expect(mockLoadingContainer).toBeInTheDocument();
  });

  it("Should render error message when error is available", async () => {
    mockUseQuery.mockImplementation(() => ({
      ...mockUseQuery,
      error: {
        message: "Error message",
      },
    }));

    const { result } = renderHook(() => useRegistrationsHook(), { wrapper });
    await waitFor(() => expect(result.current.fetch.error).not.toBeNull());

    customRender(<DashboardPage />);

    const mockErrorContainer = screen.getByTestId("error-container");

    expect(mockErrorContainer).toBeInTheDocument();
    expect(mockErrorContainer).toHaveTextContent("Error message");
  });
});
