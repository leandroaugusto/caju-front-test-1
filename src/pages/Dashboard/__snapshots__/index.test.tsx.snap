// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardPage Should show DashboardPage page 1`] = `
<div>
  <div
    class="sc-braxZu cCQSSo"
  >
    <div
      data-testid="mock-search-bar-container"
    >
      SearchBar
    </div>
    <div
      data-testid="mock-columns-container"
    >
      <div>
        3
      </div>
      <div>
        2
      </div>
    </div>
  </div>
</div>
`;
